src/uct/tcp/tcp_ep.c

```diff
@@ -257,37 +257,73 @@ unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
     return recv_length > 0;
 }
 
+static inline ucs_status_t uct_tcp_ep_am_send(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
+                                              const uct_tcp_am_hdr_t *hdr,
+                                              size_t check_length, const char *name)
+{
+    UCT_CHECK_LENGTH(hdr->length, 0, check_length, name);
+    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
+                       hdr + 1, hdr->length, "SEND fd %d", ep->fd);
+
+    ep->length         = sizeof(*hdr) + hdr->length;
+    iface->outstanding += ep->length;
+
+    uct_tcp_ep_send(ep);
+    if (ep->length > 0) {
+        uct_tcp_ep_mod_events(ep, EPOLLOUT, 0);
+    }
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_ep_am_short(uct_ep_h uct_ep, uint8_t am_id, uint64_t header,
+                                 const void *payload, unsigned length)
+{
+    uct_tcp_ep_t *ep       = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr;
+
+    UCT_CHECK_AM_ID(am_id);
+
+    if (!uct_tcp_ep_can_send(ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    hdr         = ep->buf;
+    hdr->am_id  = am_id;
+    hdr->length = sizeof(header) + length;
+
+    memcpy(hdr + 1, &header, sizeof(header));
+    memcpy((uint8_t*)(hdr + 1) + sizeof(header), payload, length);
+
+    UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, hdr->length);
+
+    return uct_tcp_ep_am_send(iface, ep, hdr,
+                              iface->config.short_size - sizeof(*hdr), "am_short");
+}
+
 ssize_t uct_tcp_ep_am_bcopy(uct_ep_h uct_ep, uint8_t am_id,
                             uct_pack_callback_t pack_cb, void *arg,
                             unsigned flags)
 {
     uct_tcp_ep_t *ep = ucs_derived_of(uct_ep, uct_tcp_ep_t);
     uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
     uct_tcp_am_hdr_t *hdr;
-    size_t packed_length;
+
+    UCT_CHECK_AM_ID(am_id);
 
     if (!uct_tcp_ep_can_send(ep)) {
         return UCS_ERR_NO_RESOURCE;
     }
 
     hdr         = ep->buf;
     hdr->am_id  = am_id;
-    hdr->length = packed_length = pack_cb(hdr + 1, arg);
-    ep->length  = sizeof(*hdr) + packed_length;
+    hdr->length = pack_cb(hdr + 1, arg);
 
-    UCT_CHECK_LENGTH(hdr->length, 0,
-                     iface->config.buf_size - sizeof(uct_tcp_am_hdr_t),
-                     "am_bcopy");
     UCT_TL_EP_STAT_OP(&ep->super, AM, BCOPY, hdr->length);
-    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
-                       hdr + 1, hdr->length, "SEND fd %d", ep->fd);
-    iface->outstanding += ep->length;
 
-    uct_tcp_ep_send(ep);
-    if (ep->length > 0) {
-        uct_tcp_ep_mod_events(ep, EPOLLOUT, 0);
-    }
-    return packed_length;
+    uct_tcp_ep_am_send(iface, ep, hdr,
```

Comment:
`uct_tcp_ep_am_send` can fail - need to check status. (also stats needs to be increased in case of success only)